goldenmaster/apigear/mqtt/private/mqttcwrapper.cpp

```diff
@@ -0,0 +1,587 @@
+    }).detach();
+}
+
+void CWrapper::handleTextMessage(const Message& message)
+{
+    AG_LOG_INFO("new msg: topic " + message.topic + " msg content: " + message.content);
+    auto subscribedTopicsRange = m_subscribedTopics.equal_range(message.topic);
+    std::pair<std::multimap<std::string, CallbackFunction>::iterator, std::multimap<std::string, CallbackFunction>::iterator> topics = subscribedTopicsRange;
+    for (auto iter = topics.first; iter != topics.second; ++iter)
+    {
+        if(iter->second != nullptr)
+        {
+            iter->second(message.topic, message.content, message.responseTopic, message.correlationData);
+        }
+    }
+}
+
+/**
+ * @brief helper function to wrap the c call
+ * 
+ * @param options reference to the options which where the topic should be added
+ * @param responseTopic the topic which is set as response topic
+ */
+void MQTTProperties_addResponseTopic(MQTTAsync_responseOptions& options, const Topic& responseTopic)
+{
+    MQTTProperty responseTopicProperty;
+    responseTopicProperty.identifier = MQTTPROPERTY_CODE_RESPONSE_TOPIC;
+    responseTopicProperty.value.data = { static_cast<int>(responseTopic.getEncodedTopic().size()), const_cast<char*>(responseTopic.getEncodedTopic().c_str()) };
+    MQTTProperties_add(&(options.properties), &responseTopicProperty);
+}
+
+/**
+ * @brief helper function to wrap the c call
+ * 
+ * @param options reference to the options which where the topic should be added
+ * @param responseId the Id which is used as correlation data
+ */
+void MQTTProperties_addResponseIdAsCorrData(MQTTAsync_responseOptions& options, int responseId)
+{
+    MQTTProperty correlationDataProperty;
+    correlationDataProperty.identifier = MQTTPROPERTY_CODE_CORRELATION_DATA;
+    const std::string correlationData = { std::to_string(responseId) } ;
+    correlationDataProperty.value.data = { static_cast<int>(correlationData.size()), const_cast<char*>(correlationData.c_str()) };
+    MQTTProperties_add(&(options.properties), &correlationDataProperty);
+
+}
+
+void CWrapper::invokeRemote(const Topic& topic, const Topic& responseTopic, const std::string& value, int responseId)
+{
+    MQTTAsync_responseOptions opts = MQTTAsync_responseOptions_initializer;
+    MQTTAsync_message pubmsg = MQTTAsync_message_initializer;
+
+    MQTTProperties_addResponseTopic(opts, responseTopic);
+    MQTTProperties_addResponseIdAsCorrData(opts, responseId);
+
+    opts.onFailure5 = onSendFailure;
+    opts.context = new genericContext{getPtr()};
+    pubmsg.payload = const_cast<void*>(static_cast<const void*>(value.c_str()));
+    pubmsg.payloadlen = static_cast<int>(value.size());
+    pubmsg.qos = QOS;
+    pubmsg.retained = 0;
+
+    // the responseOptions properties do get overwritten by the msg properties later
+    pubmsg.properties = opts.properties;
+
+    sendMessage(topic.getEncodedTopic(), &pubmsg, &opts);
+}
+
+void CWrapper::notifyPropertyChange(const Topic& topic, const std::string& value)
```

Comment:
instead of this messages I'd expose functions like
For simple ones
sendMessage(topic, payload) and sendRetainedMessage(topic, payload)  or sendMessage(topic, payload, bool)
{
    MQTTAsync_responseOptions opts = MQTTAsync_responseOptions_initializer;
    MQTTAsync_message pubmsg = MQTTAsync_message_initializer;

    opts.onFailure5 = onSendFailure; // Not sure about this line and one below it is in Property Change, but not in  notifySignal, setRemoteProperty. Is it necessary for Property Change
    opts.context = new genericContext{getPtr()}; 
    pubmsg.payload = const_cast<void*>(static_cast<const void*>(value.c_str()));
    pubmsg.payloadlen = static_cast<int>(value.size());
    pubmsg.qos = QOS;
    // property changes shall be retained and automatically send to new clients
    pubmsg.retained = 1;

    sendMessage(topic.getEncodedTopic(), &pubmsg, &opts);
}
and sendMessage(..., context)
which requires also createContextFunction sth like
createOptions and then add(opts, identifier, data)
{
    MQTTProperty dataProperty;
    correlationDataProperty.identifier = identifier;
    correlationDataProperty.value.data = { static_cast<int>(value)), const_cast<char*>(value.c_str()) };
    MQTTProperties_add(&(opts.properties), &correlationDataProperty);
}
this way the wrapper doesn't know about olink protocol and client and servers side add their specific logic, also some repetitive code is removed

Reply:
Created issue for this https://github.com/apigear-io/template-cpp14/issues/108